odxtools/cli/decode.py

```diff
@@ -0,0 +1,128 @@
+# SPDX-License-Identifier: MIT
+import argparse
+from typing import Dict, List, Optional
+
+from ..database import Database
+from ..diagservice import DiagService
+from ..exceptions import odxraise
+from ..odxtypes import ParameterValue
+from ..singleecujob import SingleEcuJob
+from . import _parser_utils
+
+# name of the tool
+_odxtools_tool_name_ = "decode"
+
+
+def get_display_value(v: ParameterValue) -> str:
+    if isinstance(v, bytes):
+        return v.hex(" ")
+    elif isinstance(v, int):
+        return f"{v} (0x{v:x})"
+    else:
+        return str(v)
+
+
+def print_summary(
+    odxdb: Database,
+    ecu_variants: Optional[List[str]] = None,
+    data: bytes = b'',
+    decode: bool = False,
+) -> None:
+    ecu_names = ecu_variants if ecu_variants else [ecu.short_name for ecu in odxdb.ecus]
+    services: Dict[DiagService, List[str]] = {}
+    for ecu_name in ecu_names:
+        ecu = odxdb.ecus[ecu_name]
+        if not ecu:
+            print(f"The ecu variant '{ecu_name}' could not be found!")
+            continue
+        if data:
+            found_services = ecu._find_services_for_uds(data)
+            for found_service in found_services:
+                ecu_names = services.get(found_service, [])
+                ecu_names.append(ecu_name)
+                services[found_service] = ecu_names
+
+    print(f"Binary data: {data.hex(' ')}")
+    for service, ecu_names in services.items():
+        if isinstance(service, DiagService):
+            print(
+                f"Decoded by service '{service.short_name}' (decoding ECUs: {', '.join(ecu_names)})"
+            )
+        elif isinstance(service, SingleEcuJob):
+            print(
+                f"Decoded by single ecu job '{service.short_name}' (decoding ECUs: {', '.join(ecu_names)})"
+            )
+        else:
+            print(f"Decoded by unknown diagnostic communication: '{service.short_name}' "
+                  f"(decoding ECUs: {', '.join(ecu_names)})")
+
+        if decode:
+            if data is None:
+                odxraise("Data is required for decoding")
+
+            decoded = service.decode_message(data)
+            print(f"Decoded data:")
+            for param_name, param_value in decoded.param_dict.items():
+                print(f"  {param_name}={get_display_value(param_value)}")
+
+
+def add_subparser(subparsers: "argparse._SubParsersAction") -> None:
+    parser = subparsers.add_parser(
+        "decode",
+        description="\n".join([
+            "Decode request by hex-data",
+            "",
+            "Examples:",
+            "  For displaying the service associated with the request 10 01 & decoding it:",
+            "    odxtools decode ./path/to/database.pdx -d 10 01",
```

Comment:
```suggestion
            "    odxtools decode ./path/to/database.pdx -d '10 01'",
```

Reply:
oops, yes, thanks, i fixed it to work with -d 10 01 again. -D is a modifier to additionally decode it, although maybe it might be better to just differentiate between -d 10 01 and -D 10 01 - thoughts?

Reply:
allowing the data to be specified as multiple arguments is a bad idea IMO, see below...